Update README.md (#190)
neuralmagic/sparsify: README.md

    </a>
</p>

## Overview

Use Sparsify to easily create sparse, performant models. Sparsify can help you optimize models from scratch or sparse transfer learn onto your own data to target best-in-class inference performance on your target deployment hardware. To get started with the Sparsify Cloud UI, create an account [here](https://app.neuralmagic.com) and check out the [Sparsify Quickstart Guide](https://docs.neuralmagic.com/sparsify/quickstart).

Sparsify makes applying state-of-the-art [sparsification](https://docs.neuralmagic.com/main/source/getstarted.html#sparsification) algorithms using techniques such as pruning and quantization to any neural network easy with a simple UI and one-command API calls that removes the complexity of hyperparameter tuning and sparsification targets. 

> 🚨**Note**🚨: Sparsify is currently in an alpha state meaning that many of the pathways and underlying functionaliity may not be fully stable. Please provide your feedback on your experience, what you do not like about the Sparsify Local APIs and what you'd like to see next. Your feedback will be integral in driving the future of Sparsify forward and we thank you in advance. Provide feedback (bug reports, UI issues, CLI errors, or just say hello to our team) [here](https://neuralmagic.com/support/). 

Reviewer: Do you want to direct them straight to the sparsify GitHub issue queue?

Author: Yeah, That's probably more appropriate for this alpha if you think that the support link will get overfilled.

> 🚨**Note**🚨: Sparsify is currently in an alpha state meaning that many of the pathways and underlying functionaliity may not be fully stable. Please provide your feedback on your experience, what you do not like about the Sparsify Local APIs and what you'd like to see next. Your feedback will be integral in driving the future of Sparsify forward and we thank you in advance. Provide feedback (bug reports, UI issues, CLI errors, or just say hello to our team) [here](https://neuralmagic.com/support/). 

Reviewer: ```suggestion
🚨**Note**🚨: Sparsify is currently in an alpha state meaning that many of the pathways and underlying functionality may not be fully stable. Provide your feedback on your experience, what you do not like about the Sparsify Local APIs, and what you'd like to see next. Your feedback will be integral in driving the future of Sparsify forward and we thank you in advance. Provide feedback (bug reports, UI issues, CLI errors, or just say hello to our team) [here](https://neuralmagic.com/support/). 
```

 ## Recommended Hardware Support and System Requirements

> Sparsify requires the same hardware that is required to train standard Neural Networks such as a GPU, TPU, or other specialized hardware for training. We recommend you use a Linux environment with a GPU that has a minimum of X RAM, X GPU Memory, X number of cores and is CUDA-enabled. If you encounter issues setting up your training environment, please file a GitHub issue [here]( 

Reviewer: ```suggestion
Sparsify requires the same hardware that is required to train standard neural networks such as a GPU, TPU, or other specialized hardware for training. We recommend you use a Linux environment with a GPU that has a minimum of X RAM, X GPU Memory, X number of cores, and is CUDA-enabled. If you encounter issues setting up your training environment, file a GitHub issue [here]( https://github.com/neuralmagic/sparsify/issues).
```

https://github.com/neuralmagic/sparsify/issues). 

Sparsify is tested on Python 3.7-3.10, ONNX 1.5.0-1.12.0, ONNX opset version 11+, and manylinux compliant systems. Using a  [virtual environment](https://docs.python.org/3/library/venv.html)  is highly recommended.

## Available Sparsify APIs

> To run the Sparsify APIs locally, all you need to do is create an account, install and login to Sparsify via your CLI, and provide the required arguments for the experment type you wish to run outlined below. All Sparsify API calls start with the `sparsify.run `command amnd have the following arguments in aggregate: 

Reviewer: ```suggestion
To run the Sparsify APIs locally, all you need to do is create an account, install and log in to Sparsify via your CLI, and provide the required arguments for the experiment type you wish to run outlined below. All Sparsify API calls start with the `sparsify.run` command and have the following arguments in aggregate: 
```

```bash
sparsify.run experiment-type \  
--model {MODEL_PATH} \  
--data {DATA_PATH} \  
--use-case use-case \  
--optim-level optim-level
```

> Sparsify has 3 experiment types which contain their own set of required arguments: 

Reviewer: ```suggestion
Sparsify has three experiment types which contain their own set of required arguments: 
```

```bash
- sparsify.run one-shot ...
- sparsify.run sparse-transfer ...
- sparsify.run training-aware ...
```

> - **sparsify. run one-shot** enables you to provide a dense model + calibration dataset and run the Neural Magic One-Shot API on it to generate a sparse model, transfer learned on your calibration dataset all without a full retraining of the model; reducing your time to deploy. The Sparsify One-Shot API creates a sparse ONNX model in the working directory. 

Reviewer: ```suggestion
- **sparsify. run one-shot** enables you to provide a dense model and calibration dataset and run the Neural Magic One-Shot API on it to generate a sparse model, transfer learned on your calibration dataset all without full retraining of the model; reducing your time to deploy. The Sparsify One-Shot API creates a sparse ONNX model in the working directory. 
```

> - **sparsify. run training-aware** enables you to provide a use case + training dataset (and optional dense model) and generate a sparse model on your data.  You can customize a recipe further (simple with handful of hyperparameters, advanced entire algorithms), and run training through your pipeline, package it, and deploy it. The Sparsify Training-Aware API creates a sparse ONNX model in the working directory. 

Reviewer: ```suggestion
- **sparsify. run training-aware** enables you to provide a use case and training dataset (and optional dense model) and generate a sparse model on your data.  You can customize a recipe further (simply with a handful of hyperparameters, advanced entire algorithms), and run training through your pipeline, package it, and deploy it. The Sparsify Training-Aware API creates a sparse ONNX model in the working directory. 
```

> - **sparsify. run sparse-transfer** enables you to provide a use case + training dataset and generate a sparse model via sparse transfer learning on top of an already sparse model provided by Neural Magic.  The Sparsify Sparse-Transfer API takes care of the transfer learning process and fits your data to a model optimized on your metric of choice; making generating performant models for your specific use case easy. The Sparsify Sparse-Transfer API creates a sparse ONNX model in the working directory. 

Reviewer: ```suggestion
- **sparsify. run sparse-transfer** enables you to provide a use case and training dataset and generate a sparse model via sparse transfer learning on top of an already sparse model provided by Neural Magic.  The Sparsify Sparse-Transfer API takes care of the transfer learning process and fits your data to a model optimized on your metric of choice; making generating performant models for your specific use case easy. The Sparsify Sparse-Transfer API creates a sparse ONNX model in the working directory. 
```

## Sparsify APIs Usage and Parameters 
#### Training Information: 
The Sparsify APIs will kick off the creation of an initial model based on the optimizing and satisficing metrics defaulted or selected; followed by automatic hyperparameter tuning to optimize the given metric to its highest value.  

The Sparsify training process will run locally on your training hardware and automatically adjust training commands to fit the available training hardware (adjusting gradient accumulation for that setup).  

#### Model Selection
For the One-Shot and Training-Aware APIs, you will need to provide a dense model you wish to optimize for inference. When running the `sparsify.run` command for either of these APIs you'll need to supply the argument `--model PATH_TO_DENSE_ONNX_MODEL` with the path to your model in your local working directory. The maximum model size is 2GB.
  
#### Data Selection
For all Sparsify APIs, you will need to provide a training dataset to fit to your sparse model you wish to create. You will need to make sure that your data is properly formatted to work with the Sparsify APIs. Depending on your use case, your data may need to take a specific form.  

> To view how to properly format your data for use with Sparsify, view [Neural Magic's Data Format Guide](www.asdfas/com).  Once your data is properly formatted, run the `sparsify.run` command andbe sure to supply the argument `--data PATH_TO_DATA` with the path to your data in your local working directory. 

Reviewer: ```suggestion
To view how to properly format your data for use with Sparsify, view [Neural Magic's Data Format Guide](www.asdfas/com).  Once your data is properly formatted, run the `sparsify.run` command and be sure to supply the argument `--data PATH_TO_DATA` with the path to your data in your local working directory. 
```

#### Use Case Selection:

> To use the Sparsify APIs, you must choose a use case and provide a dataset that you wish to get a sparse ONNX model for. You are required to designate both the use-case as well as provide a training dataset that will be applied for the Sparse-Transfer and Training-Aware Experiment APIs. For the One-Shot Experiment API, you'll additionally need to provide a dense model you wish to optimize for inference.

Reviewer: ```suggestion
To use the Sparsify APIs, you must choose a use case and provide a dataset that you wish to get a sparse ONNX model for. You are required to designate both the use-case as well as provide a training dataset that will be applied for the Sparse-Transfer and Training-Aware Experiment APIs. For the One-Shot Experiment API, you will additionally need to provide a dense model you wish to optimize for inference.
```

   
   **Use Cases**

>    You can select from the following lists of supported list of use-cases (--use-case) to create a performant model for: 

Reviewer: ^ does this need to be rewritten?! ...from the following list of supported user cases?

-   CV classification: [*ic, image-classification, image_classification, classification*]
-   CV detection: [*od, object-detection, object_detection, detection*]
-   CV segmentation: [*segmentation*]
-   NLP question answering: [*qa, question-answering, question_answering*]
-   NLP text classification: [*text-classification, text_classification, glue*]
-   Sentiment analysis: [*sentiment, sentiment_analysis, sentiment-analysis*]
-   NLP token classification: [*token-classification, token_classification*]
-   Named entity recognition: [*ner, named-entity-recognition, named_entity_recognition*]

#### Optimization Level:

> When using the Sparsify APIs, you will have the ability to set a model optimization level. This optimization level is a value between 0.0 and 1.0 where 0.0 is fully optimized for accuracy and 1.0 is fully optimized for performance. Note that if you fully optimize for one extreme that does not mean that there will be no sparsity present or accuracy will be completely ignored. It just means that Sparsify will optimize your model  to that metric as much as posssible, while maintaining a reasonable accuracy or performance value. 

Reviewer: ```suggestion
When using the Sparsify APIs, you will have the ability to set a model optimization level. This optimization level is a value between 0.0 and 1.0 where 0.0 is fully optimized for accuracy and 1.0 is fully optimized for performance. Note that if you fully optimize for one extreme that does not mean that there will be no sparsity present or accuracy will be completely ignored. It just means that Sparsify will optimize your model to that metric as much as possible, while maintaining a reasonable accuracy or performance value. 
```

## Overview

Sparsify enables you to accelerate inference without sacrificing accuracy by applying state-of-the-art pruning, quantization, and distillation algorithms to Neural Networks with a simple web app and one-command API calls. 

To empower you in compressing models, Sparsify is made up of two components: the Sparsify Cloud and the Sparsify CLI/API.
The Sparsify Cloud is a web application that allows you to create and manage Sparsify Experiments, explore hyperparameters, predict performance, and compare results across both Experiments and deployment scenarios.
The Sparsify CLI/API is a Python package that allows you to run Sparsify Experiments locally, sync with the Sparsify Cloud, and integrate into your own workflows.

To get started immediately, [create an account](https://account.neuralmagic.com/signup) and then check out the [Installation](https://github.com/neuralmagic/sparsify/edit/Sparsify-Alpha-README/README.md#installation) and [Quick Start](https://github.com/neuralmagic/sparsify/edit/Sparsify-Alpha-README/README.md#quick-start) sections of this README.
With all of that setup, sparsifying your models is as easy as:

```bash

> sparsify.run one-shot --use-case image_classification --model resnet50 --data imagenette --optim-level 0.5

Reviewer 2: ```suggestion
sparsify.run sparse-transfer --use-case image-classification --data imagenette --optim-level 50 --train-kwargs '{"dataset": "imagenette"}'
```
